listenbrainz/webserver/static/js/jsx/importer.test.js

```diff
@@ -0,0 +1,310 @@
+
+    expect(window.fetch).toHaveBeenCalledWith(`${props.lastfm_api_url}?method=user.getrecenttracks&user=${lastfmUsername}&api_key=${props.lastfm_api_key}&from=1&page=1&format=json`);
+  })
+
+  it('should call encodeScrobbles', async () => {
+    // Mock function for encodeScrobbles
+    importer.encodeScrobbles = jest.fn(data => ["foo", "bar"]);
+
+    const data = await importer.getPage(1);
+    expect(importer.encodeScrobbles).toHaveBeenCalledTimes(1);
+    expect(data).toEqual(["foo", "bar"]);
+  })
+
+  it('should retry if 50x error is recieved', async () => {
+    // Mock function for fetch
+    window.fetch = jest.fn().mockImplementation(() => {
+      return Promise.resolve({
+        ok: false,
+        status: 503,
+      })
+    })
+    // Mock function for console.warn
+    console.warn = jest.fn();
+
+    await importer.getPage(1);
+    // There is no direct way to check if retry has been called
+    expect(setTimeout).toHaveBeenCalledTimes(1);
+    expect(console.warn).toHaveBeenCalledWith('Got 503 fetching last.fm page=1, retrying in 3s')
+
+    jest.runAllTimers();
+  })
+
+  it('should skip the page if 40x is recieved', async () => {
+    // Mock function for failed fetch
+    window.fetch = jest.fn().mockImplementation(() => {
+      return Promise.resolve({
+        ok: false,
+        status: 404,
+      })
+    })
+
+    // Mock function for console.warn
+    console.warn = jest.fn();
+
+    await importer.getPage(1);
+    expect(console.warn).toHaveBeenCalledWith('Got 404, skipping')
+  })
+
+  it('should retry if there is any other error', async () => {
+    // Mock function for fetch
+    window.fetch = jest.fn().mockImplementation(() => {
+      return Promise.resolve({
+        ok: true,
+        json: () => Promise.reject(),
+      })
+    })
+    // Mock function for console.warn
+    console.warn = jest.fn();
+
+    await importer.getPage(1);
+    // There is no direct way to check if retry has been called
+    expect(setTimeout).toHaveBeenCalledTimes(1);
+    expect(console.warn).toHaveBeenCalledWith('Error fetching last.fm page=1, retrying in 3s')
+
+    jest.runAllTimers();
+  })
+})
+
+describe('submitPage works correctly', () => {
+  beforeEach(() => {
+    // Clear previous mocks
+    APIService.mockClear();
+
+    // Mock for getRateLimitDelay
+    importer.updateRateLimitParameters = jest.fn();
+
+    // Mock for console.warn
+    console.warn = jest.fn();
+  });
+
+  it('calls submitListens once and increments numSuccesful', async () => {
+    let spy = jest.spyOn(importer.APIService, 'submitListens').mockImplementation(async () => {
+      return {"status": 200}
+    });
+
+    importer.submitPage();
+    jest.runAllTimers();
+
+    // Flush all promises
+    // https://stackoverflow.com/questions/51126786/jest-fake-timers-with-promises
+    await new Promise(resolve => setImmediate(resolve));
+    
+    expect(importer.APIService.submitListens).toHaveBeenCalledWith("foobar", "import", undefined);
```

Comment:
Here you check if it's been called, but you could also check that it's been called only once, as your test description suggests.